templates/template.go

```diff
@@ -0,0 +1,47 @@
+package templates
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/algorand/go-algorand-sdk/types"
+)
+
+func replace(buf, newBytes []byte, offset, placeholderLength uint64) []byte {
+	output := append(buf[:offset], newBytes...)
+	return append(output, buf[(offset+placeholderLength):]...)
+}
+
+func inject(original []byte, offsets []uint64, values []interface{}) (result []byte, err error) {
+	result = original
+	if len(offsets) != len(values) {
+		err = fmt.Errorf("length of offsets %v does not match length of replacement values %v", len(offsets), len(values))
+		return
+	}
+
+	for i, value := range values {
+		decodedLength := 0
+
+		if valueAsUint, ok := value.(uint64); ok {
+			buffer := make([]byte, 8)
+			decodedLength = binary.PutUvarint(buffer, valueAsUint)
+			result = replace(result, buffer, offsets[i], uint64(decodedLength)) // TODO ejr should the last arg here just always be 1?
```

Comment:
that should not be the decoded length but the place holder length that you've left in the original `referenceAsBytes` array. I think that you have them in different lengths. Might make it more difficult for you. One byte should be sufficient.